subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java

```diff
@@ -443,7 +443,7 @@ private void onOrderingCycle(Node successor, Node currentNode) {
                 connectedNodes.addAll(taskNode.getFinalizingSuccessors());
             }
         });
-        graphWalker.add(entryNodes);
+        graphWalker.add(successor);
```

Comment:
By using the entry nodes, we miss everything which is "after" the entry nodes, so the finalizers of the entry nodes, correct? This is why we start searching for cycles from the successor. That means that any cycle which involves dependencies of the entry nodes and the finalizer would already have been detected, right?

We only did miss cycles which are introduced by the finalizer alone. Can we add a test where the finalizer is in a non-trivial cycle which is now detected and has not been detected before?

Reply:
> By using the entry nodes, we miss everything which is "after" the entry nodes, so the finalizers of the entry nodes, correct? 

We miss not only what is directly "after" the entry nodes, but any branches to finalizers within the graph if tasks from those branches don't happen to be part of the graph via another path.
I added a test showing this: https://github.com/gradle/gradle/pull/14965/commits/b88456f3f2ca392dac71384b4a388fdafdbcb989
Here `a dependsOn b dependsOn c`, `d dependsOn f dependsOn e dependsOn d (cycle)`. The two graphs are connected using a finalizer - `b finalizedBy d`. We detect that those graphs are connected and that there is a cycle as we compute the execution plan for entry task `a`. The cycle is detected when visiting `e` (currentNode) and encountering `d` (successor, already seen). When reporting the cycle, if we walk the graph following the dependsOn relationships, we only see `a -> b -> c` and the `d -> f -> e -> d` part is missed completely.
When we walk the graph from the successor `d`, we visit the cyclic part as well as `d -> b -> c`, because b is a finalizingSuccessor of d. We do not visit `a` in this case, but it is not part of the graph that has a cycle which we want to report.

> We only did miss cycles which are introduced by the finalizer alone. Can we add a test where the finalizer is in a non-trivial cycle which is now detected and has not been detected before?

With this change the existing cycle detecting tests remain green, the ones that were ignored for finalizedBy case, reproducers of related issue also turn green and I added a couple of new ones. Not sure if this last one https://github.com/gradle/gradle/pull/14965/commits/b88456f3f2ca392dac71384b4a388fdafdbcb989 is non-trivial enough, but it does show how we miss a whole part of the graph if we start at the original start node and not the offending node.